persistence/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedActorOf.scala

```diff
@@ -0,0 +1,135 @@
+package com.evolutiongaming.akkaeffect.persistence
+
+import akka.actor.Actor
+import akka.persistence.SnapshotSelectionCriteria
+
+import cats.effect.implicits.effectResourceOps
+import cats.effect.{Async, Resource, Ref}
+import cats.syntax.all._
+
+import com.evolutiongaming.akkaeffect._
+import com.evolutiongaming.catshelper.ToFuture
+
+import scala.reflect.ClassTag
+import java.time.Instant
+
+object EventSourcedActorOf {
+
+  /** Describes lifecycle of entity with regards to event sourcing & PersistentActor Lifecycle phases:
+    *
+    *   1. RecoveryStarted: we have id in place and can decide whether we should continue with recovery 
+    *   2. Recovering : reading snapshot and replaying events 
+    *   3. Receiving : receiving commands and potentially storing events & snapshots 
+    *   4. Termination : triggers all release hooks of allocated resources within previous phases
+    *
+    * @tparam S snapshot
+    * @tparam E event
+    * @tparam C command
+    */
+  type Lifecycle[F[_], S, E, C] =
+    Resource[F, RecoveryStarted[F, S, E, Receive[F, Envelope[C], ActorOf.Stop]]]
+
+  def actor[F[_]: Async: ToFuture, S, E, C: ClassTag](
+    eventSourcedOf: EventSourcedOf[F, Lifecycle[F, S, E, C]],
+    persistence: EventSourcedPersistence[F]
+  ): Actor = ActorOf[F] { actorCtx =>
+    for {
+      eventSourced    <- eventSourcedOf(actorCtx).toResource
+      recoveryStarted <- eventSourced.value
+
+      snapshotStore <- persistence.snapshotStore[S](eventSourced).toResource
+      eventStore    <- persistence.eventStore[E](eventSourced).toResource
+
+      snapshot <- snapshotStore.latest.toResource
+
+      recovering <- recoveryStarted(
+        snapshot.map(_.metadata.seqNr).getOrElse(SeqNr.Min),
+        snapshot.map(_.asOffer)
+      )
+
+      seqNr <- recovering.replay.use { replay =>
+        val snapSeqNr = snapshot.map(_.metadata.seqNr).getOrElse(SeqNr.Min)
+        val fromSeqNr = snapshot.map(_.metadata.seqNr + 1).getOrElse(SeqNr.Min)
+        for {
+          events <- eventStore.events(fromSeqNr)
+          seqNrL <- events.foldWhileM(snapSeqNr) {
+            case (_, EventStore.Event(event, seqNr)) => replay(event, seqNr).as(seqNr.asLeft[Unit])
+            case (_, EventStore.HighestSeqNr(seqNr)) => seqNr.asLeft[Unit].pure[F]
+          }
+          seqNr <- seqNrL
+            .as(new IllegalStateException("should newer happened"))
+            .swap
+            .liftTo[F]
```

Comment:
This might be a little more clear, if `match...case` was used instead of `swap/liftOf` pair. It is a matter of preference though, so feel free to ignore:
```          seqNr <- seqNrL match {
            case Left(seqNr) => seqNr.pure[F]
            case Right(()) => new IllegalStateException("should newer happened").raiseError[F, SeqNr]
          }
```

Reply:
`newer` -> `never`

Reply:
It might be worth to mention why it never happens in a comment, i.e. because `foldWhileM` only returns `Right(...)` if the block inside of it returns `Right(...)` and shortcuts the computation.